eip-0019.md

```diff
@@ -0,0 +1,238 @@
+
+1) A user creates a transaction in the Hot Wallet application and then
+presses a "Sign With Cold Wallet" button
+2) The Hot Wallet application shows a QR code with serialized transaction
+bytes on the screen
+3) User scans the QR code using the Cold Wallet device, signs the
+transaction after which QR code of the signed transaction is displayed.
+4) Then user scans the QR code of the signed transaction and sends it to blockchain.
+
+The design of Ergo contracts allows for a simple and universal implementation
+which we describe in [Reference Implementation](#reference-implementation)
+section.
+
+In the following sections we:
+ 1) describe the main problem we need to solve; 
+ 2) describe a solution;  
+ 3) specify a protocol with two roles: HotWallet and
+ColdWallet which must be implemented by complying Wallet applications and 
+ 4) describe a reference implementation of both Hot and Cold roles in [Ergo Wallet
+for Android](https://github.com/MrStahlfelge/ergo-wallet-android).
+
+## Cold Signing Problem
+
+In the Ergo's eUTXO model a box can be protected by an arbitrary complex
+contract (aka spending condition) and the spending transaction should satisfy
+that condition by adding required context variables, creating expected number of
+outputs with specific registers etc. i.e. a special data structure called
+`Context`. The Context should be created for each input of the transaction
+and then passed to the Prover which will generate a signature for that input.
+See [general overview of signing and
+verification](https://github.com/ScorexFoundation/sigmastate-interpreter#sigma-language-background)
+process in Ergo for details.
+
+In general, the Context represents the current state of the blockchain and
+includes current header, previous 10 headers, current height etc. This data can
+be retrieved from blockchain nodes. This is possible on Hot Wallet device, but
+is not possible on Cold Wallet device (there is no network connection).
+
+At the same time the prover need to know both the Context data and the private
+keys, which are stored on the Cold Wallet device, and so the Prover must run on
+the Cold Wallet device.
+
+And finally, which is the problem, we cannot transfer unsigned transaction along
+with all the contexts for each input to the Cold Wallet via QR code. 
+
+QR codes have limit of 4K bytes on the maximum size of serialized data. Most of
+the transactions with required Contexts will exceed this limit.
+
+## Simplified Signing using ReducedTransaction
+
+To solve the problem of _cold signing_ we need a new data structure and
+serialization format called `ReducedTransaction`.
+
+```
+ReducedTransaction:
+  - unsignedTx: UnsignedTransaction
+  - reducedInputs: Seq[ReducedInputData]
+
+UnsignedTransaction:
+  - inputs: Seq[UnsignedInput],
+  - dataInputs: Seq[DataInput],
+  - outputCandidates: Seq[ReductionResult]
+
+UnsignedInput:
+  - boxId: BoxId
+  - extension: ContextExtension
+```
+
```

Comment:
```suggestion

ReducedInputData:
  - reductionResult: ReductionResult
  - contextExtension: ContextExtension
  
```
based on appkit implementation

Reply:
There is one subtlety here. ReducedInputData indeed has ContextExtension, however it is not serialized because it is already serialized as part of Input.